### DISCUSSION TITLE
Component Data Interfaces for Vue 2?

### DISCUSSION THREAD

**Discussion by douglasg14b** [vuejs/language-tools#1016] [Q&A] [answered]
It's mentioned that we have these here: https://github.com/johnsoncodehk/volar/issues/418

However, I can't seem to get these working with Vue 2. Do we not get auto-completion & docs with Vue 2 & volar?

Alternatively, for Vue 2, we already have JSON bits that define the APIs we all have. Any reason that cannot be supported for compatibility?

<hr>

The example here: https://github.com/johnsoncodehk/volar/tree/master/extensions/vscode-vue-language-features

Doesn't seem to work in Vue2?

```ts
// components.d.ts
declare module '@vue/runtime-core' {
  export interface GlobalComponents {
    RouterLink: typeof import('vue-router')['RouterLink']
    RouterView: typeof import('vue-router')['RouterView']
  }
}

export {}
```

I still get no autocomplete or API info for router links or views.

Are there more explicit exmaples of defining components this way that we can use? Especially for component libraries that are not using typescript?

**Comment by xiaoxiangmoe** [comment, marked as answer]
Because vue-router 3(vue 2) has no types for it. It only works in vue-router 4(vue 3).

**Reply by douglasg14b** [reply]
Thanks for explaining the example!

@johnsoncodehk FYI I have un-marked this as the answer, as that is an good explanation for the example, but not an answer to the question:

> Are there more explicit examples of defining components this way that we can use? Especially for component libraries that are not using typescript?

So in this case, Vue Router 3, Vuetify 2....etc

Since Vue 2 projects will be kind of forced to move to Volar to some degree (If they want decent TS support, Pinia...etc), not having any sort of data interfaces for components is going to become a real problem. How can this be done, should this be a feature request?

**Reply by xiaoxiangmoe** [reply]
@douglasg14b 
I'm trying to add types to vue-router 3. You can track it in  https://github.com/vuejs/vue-router/pull/3733

If this PR was merged and released, you can use it as vue-router 4.

**Reply by douglasg14b** [reply]
Oh that's a wonderful example of what is needed!

Thanks

**Comment by johnsoncodehk** [comment]
We already supported language service plugin: https://github.com/johnsoncodehk/volar/discussions/1027

You can implement most of vetur specific features with plugins such as json base component intellisense, customizable scaffold snippets, ESLint...

I still have a lot of backlog on volar, so I don't plan to do it. If you want to do it, you can contact me on the discord server and I'm happy to help.

**Reply by douglasg14b** [reply]
>  json base component intellisense

Wait, so with a plugin I could implement support for the existing Vetur JSON definition bits? If so, then yes! I would love to, but ofc I have absolutely no idea where to even start on that....

When I get a chance I'll swing back around to this (We want Volar), and see about asking some questions to get me starting down that path.
